### DISCUSSION TITLE
Question about MultiplexForecaster use

### DISCUSSION THREAD

**Discussion by miraep8** [alan-turing-institute/sktime#2514] [Q&A] [answered]
Hey all!

I have been working a bit on implementing a `|` dunder method for `MultiplexForecaster` and thinking about how to best implement `MultiplexTransformer` as part of  #2459.  In the process of doing so I kinda went down a bit of a rabbit hole/had a few confusing things crop up - so just wanted to check in to make sure that I am being consistent with how you guys think about this class: 

#### Setup: 
- First -from what I can tell there didn't seem to be an explicit test function written to test consistency of `MultiplexForecaster`. (totally possible I may have missed it though!  I checked all the functions in sktime/forecasting/compose/tests and I didn't see anything there! 
 There is a `get_test_params()` classmethod defined a la #2255 but since this pre-assigns a `selected_forecaster` it didn't really seem like it would be useful for testing the full functionality of `MultiplexForecaster`) and since I also wanted to write some tests for my dunder anyway I thought it might be worth testing `MultiplexForecaster` in the process!
- My plan for doing so was essentially - "fit" `MultiplexForecaster` using `ForecastingGridSearchCV` (since that is the recommended use case) and then verify that the selected forecaster was also the "best scoring" forecaster if we simply used `evaluate` (from `sktime.forecasting.model_evaluation`) on each of the forecasters we put in MultiplexForecaster separately. 
#### Confusion:
- However, this led me to be a bit confused about the relative roles of `ForecastingGridSearchCV` and `MultiplexForecaster`. 
-  I was a bit confused by the fact that `ForecastingGridSearchCV` never seems to modify `selected_forecaster` in `MultiplexForecaster`. - ie essentially the workflow seems to be - make a `MultiplexForecaster`, pass it to `ForecastingGridSearchCV`, then actually use `ForecastingGridSearchCV` as your forecaster - ie can call `.predict()` and `.update()` on your `ForecastingGridSearchCV` object directly (eg, how it is used in #821) 
- I guess this was confusing to me for 2 reasons.  1 - I think from the names I was naively expecting `MultiplexForecaster` to behave like a forecasting object, and `ForecastingGridSearchCV` to just be a tool for "fitting"/finding the best forecaster in `MultiplexForecaster` for the task at hand, however it seems like the reality is perhaps swapped - ie `ForecastingGridSearchCV` is actually what ultimately behaves as your forecaster and `MultiplexForecaster` is simply the tool to display the different forecasters as a parameter space to search over. 
- It was also surprising because `MultiplexForecaster` seems to have a lot of the functionality in place for acting like a forecaster/some of the functions seems to act on the assumption that `selected_forecaster` is eventually set, though I could not actually see where this would happen directly. Though I should say it seems like some of that functionality (ie `fit()` will be removed if #2458 is merged!) so perhaps you have all already thought of this as well! 

Obviously I want to be consistent with what you guys already have in place, just wanted to check in that my interpretation checks out. I guess broad strokes what I am trying to ask here is - should I just think of `MultiplexForecaster` as something that is essentially a convenient forecaster-like wrapper that allows other functions (eg `ForecastingGridSearchCV`) to treat a list of forecasters as simply another forecaster parameter that can be optimized?  If so then what is the use of `selected_forecaster`? Or am I missing something here? Also - if my understanding is correct than it makes more sense why there might not be the explicit tests I was expecting haha. 

*Also - apologies in advance if Github Discussions is not the best place to start this chat!*

**Comment by aiwalter** [comment, marked as answer]
Hi @miraep8 , I think I can answer here as I co-developed `MultiplexForecaster` :)

> My plan for doing so was essentially - "fit" MultiplexForecaster using ForecastingGridSearchCV (since that is the recommended use case) and then verify that the selected forecaster was also the "best scoring" forecaster if we simply used evaluate (from sktime.forecasting.model_evaluation) on each of the forecasters we put in MultiplexForecaster separately.

That would be great!.

In general, `MultiplexForecaster` is meant to enable a meta-grid search functionality. This is actually also possible directly via `ForecastingGridSearchCV` with `sklearn` notation as shown in the docstring example. 
https://github.com/alan-turing-institute/sktime/blob/725740eaf0c36dd3d4c04496c77223c3e13a40c2/sktime/forecasting/model_selection/_tune.py#L414-L454

But this can be computationally expensive, so you might first want to find each best model separately and then find the best of the best with `MultiplexForecaster`. With `MultiplexForecaster` however you dont need to fit all of them at same time but could still benchmark them together. The use of `selected_forecaster` is simple to add a param that we can tune with a grid search to do the benchmarking by means of a grid search. I hope things became a bit more clear? :)

**Reply by miraep8** [reply]
Hey @aiwalter, 

Thanks for your response! It was helpful to get a sense of your intent of how this should be used/not be used.  Will make a PR soon with the added tests, and then another one in a bit with the dunder!

**Comment by fkiraly** [comment]
@miraep8, to answer some of your questions:

> First -from what I can tell there didn't seem to be an explicit test function written to test consistency

There is an entire suite which tests interface conformance (the test suite you get when you run the `check_estimator` utility). But, it does not test the logic.

As far as I know, there is no test which tests the logic of multiplexforecaster.

**Reply by miraep8** [reply]
Great :)  Just submitted a PR which hopefully addresses this!  Thank you though for verifying!  Yes, I saw that via #2255 it would test some things assuming `selected_forecaster` is set - but thought there would be merit in adding some testing to see that `MultiplexForecaster` works as expected with `ForecastingGridSearchCV`.

**Comment by fkiraly** [comment]
> Obviously I want to be consistent with what you guys already have in place, just wanted to check in that my interpretation checks out. I guess broad strokes what I am trying to ask here is

High-level: both the `MultiplexForecaster` and `GridSearchCV` *are* forecasters!
Especially the point about grid search is frequently confusing, since the "naive" implementation indeed "acts" on a forecaster and sets its parameters. The subtle point is that (a) if naively implemented, this could easily lead to "training on the test set", and (b) you don´t have an object that corresponds to the algorithm "tune your forecaster, and then apply it" (which formally is a forecatser).

Conceptually, this is nothing new above scikit-learn - the `GridSearchCV` of scikit-learn *also* acts as a forecaster, but people commonly confuse it and, in their confusion, might not use it as supervised learner, but only to "get the best estimator" from the inside.

> should I just think of MultiplexForecaster as something that is essentially a convenient forecaster-like wrapper that allows other functions (eg ForecastingGridSearchCV) to treat a list of forecasters as simply another forecaster parameter that can be optimized?

Correct!

> If so then what is the use of selected_forecaster? 

In you words, it is the "forecaster parameter that can be optimized"

> ie `ForecastingGridSearchCV` is actually what ultimately behaves as your forecaster and `MultiplexForecaster` is simply the tool to display the different forecasters as a parameter space to search over.

Yes and no - both `MultiplexForecaster` and  `ForecastingGridSearchCV` *are forecasters*!

`MultiplexForecaster` just encodes the behaviour of "act as one of the forecasters on the list" (and which can be set using the parameter).
It is the estimator counterpart of, and named after, the multiplexer (or mux) switch element https://en.wikipedia.org/wiki/Multiplexer.
I.e., it just encodes the switching behaviour, but still behaves like a forecaster.

Why `ForecastingGridSearchCV` is also a forecaster was discussed above.

In combination, both together allow you to perform what some people call "autoML", i.e., hyper-parameter tuning where the models themselves become hyper-parameters.

**Reply by miraep8** [reply]
Makes sense!  Reading @aiwalter 's description, and of course reading the code/also the documentation in more depth I came to this conclusion myself as well :)  I think sadly I did indeed fall into the camp of people who was initially confused about `GridSearchCV` being a forecaster which is where a lot of my questions stemmed from. Thank you both for taking the time to help me understand!

**Reply by fkiraly** [reply]
>  I think sadly I did indeed fall into the camp of people who was initially confused about GridSearchCV being a forecaster

But that´s, in a sense, the natural conceptualization here. Not a mistake in itself.

Making the `GridSearchCV` a forecaster (or supervised learning in sklearn) turns out to be the better idea, but it´s not the obvious one! (it´s also not my idea)

**Reply by fkiraly** [reply]
I guess that could be clearer in the documentation...
